op-service/rethdb-reader/src/lib.rs

```diff
@@ -15,19 +16,76 @@ pub unsafe extern "C" fn rdb_read_receipts(
     block_hash: *const u8,
     block_hash_len: usize,
     db_path: *const c_char,
-) -> ReceiptsResult {
-    read_receipts_inner(block_hash, block_hash_len, db_path).unwrap_or(ReceiptsResult::fail())
+) -> *mut ReceiptsResult {
+    let res = read_receipts_inner(block_hash, block_hash_len, db_path);
+    match res {
+        Ok(res) => Box::into_raw(Box::new(res)),
+        Err(err) => Box::into_raw(Box::new(ReceiptsResult::fail(err.to_string()))),
+    }
```

Comment:
Can we do this functionally instead? Should also be able to avoid allocating more onto the heap in the happy path.
```suggestion
    read_receipts_inner(block_hash, block_hash_len, db_path)
        .as_mut()
        .map(|r| r as *mut ReceiptsResult)
        .unwrap_or_else(|e| Box::into_raw(Box::new(ReceiptsResult::fail(e.to_string()))))

```

Reply:
We actually do want to allocate on both paths since we're returning an "opaque object", i.e. a `* ReceiptsResult`. The user will then free the `* ReceiptsResult` rather than the `data` inside once they're done using it.